tools-and-tests/simulator/src/main/java/org/hiero/block/simulator/mode/impl/PublishClientManager.java

```
@@ -0,0 +1,113 @@
+    }
+
+    public void init() {
+        blockStreamManager.init();
+        currentClient.init();
+    }
+
+    @Override
+    public void start() throws BlockSimulatorParsingException, IOException, InterruptedException {
+        currentHandler.start();
+    }
+
+    @Override
+    public void stop() throws InterruptedException {
+        currentHandler.stop();
+    }
+
+    private void initializeNewClientAndHandler(Block nextBlock, PublishStreamResponse publishStreamResponse) {
+        streamEnabled = new AtomicBoolean(true);
+        currentClient = new PublishStreamGrpcClientImpl(
+                grpcConfig, blockStreamConfig, metricsService, streamEnabled, startupData);
+        currentHandler =
+                new PublisherClientModeHandler(blockStreamConfig, currentClient, blockStreamManager, metricsService);
+
+        currentHandler.setPublishClientManager(this);
+        currentHandler.init();
+
+        if (nextBlock != null) {
+            long blockNumber = publishStreamResponse.getEndStream().getBlockNumber();
+            Code status = publishStreamResponse.getEndStream().getStatus();
+
+            switch (status) {
+                case TIMEOUT, BAD_BLOCK_PROOF -> blockStreamManager.resetToBlock(blockNumber - 1);
+                case BEHIND, DUPLICATE_BLOCK, SUCCESS -> blockStreamManager.resetToBlock(blockNumber + 1);
+                case INTERNAL_ERROR, PERSISTENCE_FAILED -> blockStreamManager.resetToBlock(blockNumber);
+            }
+        }
+    }
+
+    public void handleEndStream(Block nextBlock, PublishStreamResponse publishStreamResponse) {
+        try {
+            stop();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+
+        initializeNewClientAndHandler(nextBlock, publishStreamResponse);
+
+        try {
+            currentHandler.start();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to restart handler", e);
```

Catching an exception just to throw it right after that probably means we should simply propagate the exception upwards (we do not know how to deal with it here.

---

It looks like we're wrapping the exception here, which is a somewhat common approach to avoid declaring checked exceptions.
It's not my favorite pattern, but it is very popular.